MPCHomeControl/src/tools/sun.rs

```diff
@@ -0,0 +1,695 @@
+extern crate nalgebra as na;
+
+use chrono::{DateTime, Datelike, Timelike, Utc};
+use na::{Dot, Norm, Vector3};
+use uom::si::{
+    angle::{degree, radian},
+    f64::{Angle, HeatFluxDensity, Length, Pressure, Ratio, TemperatureInterval},
+    heat_flux_density::watt_per_square_meter,
+    length::centimeter,
+    pressure::pascal,
+    ratio::{percent, ratio},
+    temperature_interval::kelvin,
+};
+
+const SOLAR_CONST: f64 = 1367.0; // W/m^2
+
+// cloud model constants
+const CLEAR_SKY_MOLECULAR_SCATTERING: f64 = 0.0685;
+const LOW_CLOUDS_TRANSMITTANCE: f64 = 0.36;
+const MEDIUM_CLOUDS_TRANSMITTANCE: f64 = 0.38;
+const HIGH_CLOUDS_TRANSMITTANCE: f64 = 0.38;
+const LOW_CLOUDS_ALBEDO: f64 = 0.36;
+const MEDIUM_CLOUDS_ALBEDO: f64 = 0.38;
+const HIGH_CLOUDS_ALBEDO: f64 = 0.38;
+
+/// Get three dimensional Vector from azimuth and zenith angle.
+/// This can be used to get the norm vector of a surface or the Sun.
+///
+/// Output vector coordinate system is following:
+/// - north -> positive x axis / south -> negative x axis
+/// - east -> positive y axis / west -> negative y axis
+/// - z: up
+///
+/// # Arguments
+/// * `azimuth` - angle between north and the wall normal / sun vector
+/// * `zenith_angle` - angle between the Sun and vertical axis
+///
+/// # Returns
+/// * `Vector3<f64>` - three dimensional vector
+pub fn get_vector_from_azimuth_zenith(azimuth: &Angle, zenith_angle: &Angle) -> Vector3<f64> {
+    get_vector_from_azimuth_elevation(azimuth, &(Angle::new::<degree>(90.0) - *zenith_angle))
+}
+
+/// Get three dimensional Vector from azimuth and elevation angle.
+/// This can be used to get the norm vector of a surface or the Sun.
+///
+/// Output vector coordinate system is following:
+/// - north -> positive x axis / south -> negative x axis
+/// - east -> positive y axis / west -> negative y axis
+/// - z: up
+///
+/// # Arguments
+/// * `azimuth` - angle between north and the wall normal / sun vector
+/// * `elevation_angle` - angle between the surface and horizontal axis
+///
+/// # Returns
+/// * `Vector3<f64>` - three dimensional vector
+pub fn get_vector_from_azimuth_elevation(azimuth: &Angle, elevation_angle: &Angle) -> Vector3<f64> {
+    let x = azimuth.cos().get::<ratio>() * elevation_angle.cos().get::<ratio>();
+    let y = azimuth.sin().get::<ratio>() * elevation_angle.cos().get::<ratio>();
+    let z = elevation_angle.sin().get::<ratio>();
+    Vector3::new(x, y, z).normalize()
+}
+
+// Get a coefficient value used for calculating effectively illuminated area
+// from sun and surface azimuth and angle.
+//
+// # Arguments
+// * `surface_azimuth` - orientation of the surface
+// * `surface_angle` - angle between the surface and horizontal plane
+// * `solar_azimuth` - azimuth of the sun
+// * `solar_zenith` - zenith angle of the sun (angle between the sun vector and the z axis)
+//
+// # Returns
+// * `f64` - dot product of the surface normal and the sun vector
```

Comment:
tady porad mas ten "dot product", pritom je to spis ten "illumination coefficient"